app/models/master_file.rb

```diff
@@ -239,17 +239,23 @@ def set_workflow( workflow  = nil )
 
   # This requires the MasterFile having an actual id
   def media_object=(mo)
+    self.save!(validate: false) unless self.persisted?
+
     # Removes existing association
     if self.media_object.present?
-      self.media_object.master_files = self.media_object.master_files.to_a.reject { |mf| mf.id == self.id }
-      self.media_object.ordered_master_files = self.media_object.ordered_master_files.to_a.reject { |mf| mf.id == self.id }
-      self.media_object.save
+      self.media_object.section_ids -= [self.id]
+      self.media_object.save(validate: false)
     end
 
     self._media_object=(mo)
+    self.save!(validate: false)
+
     unless self.media_object.nil?
-      self.media_object.ordered_master_files += [self]
-      self.media_object.save
+      self.media_object.section_ids += [self.id]
+      self.media_object.save(validate: false)
```

Comment:
The saves are `validate: false` because `MediaObject` has validations that aren't necessarily show stoppers which we wouldn't want failing and breaking this method part of the way through it.  I know it didn't have it before but I think it would be safer to skip validations for this operation which should be allowed to proceed regardless of the validity of the objects.